api/grpcserver/globalstate_service.go

```diff
@@ -271,23 +271,45 @@ func (s GlobalStateService) AccountDataStream(in *pb.AccountDataStreamRequest, s
 
 	// Subscribe to the various streams
 	var (
-		channelAccount chan types.Address
-		channelReward  chan events.Reward
-		channelReceipt chan events.TxReceipt
+		accountCh       <-chan interface{}
+		rewardsCh       <-chan interface{}
+		receiptsCh      <-chan interface{}
+		accountBufFull  <-chan struct{}
+		rewardsBufFull  <-chan struct{}
+		receiptsBufFull <-chan struct{}
 	)
 	if filterAccount {
-		channelAccount = events.GetAccountChannel()
+		if accountSubscription := events.SubscribeAccount(); accountSubscription != nil {
+			defer closeSubscription(accountSubscription)
+
+			accountCh, accountBufFull = consumeEvents(stream.Context(), accountSubscription.Out())
+		}
 	}
 	if filterReward {
-		channelReward = events.GetRewardChannel()
+		if rewardsSubscription := events.SubscribeRewards(); rewardsSubscription != nil {
+			defer closeSubscription(rewardsSubscription)
+			rewardsCh, rewardsBufFull = consumeEvents(stream.Context(), rewardsSubscription.Out())
+		}
 	}
 	if filterReceipt {
-		channelReceipt = events.GetReceiptChannel()
+		if receiptsSubscription := events.SubscribeReceipts(); receiptsSubscription != nil {
+			defer closeSubscription(receiptsSubscription)
+			receiptsCh, receiptsBufFull = consumeEvents(stream.Context(), receiptsSubscription.Out())
+		}
 	}
 
 	for {
 		select {
-		case updatedAccount, ok := <-channelAccount:
+		case <-accountBufFull:
+			log.Info("account buffer is full, shutting down")
+			return fmt.Errorf("account buffer is full")
+		case <-rewardsBufFull:
+			log.Info("rewards buffer is full, shutting down")
+			return fmt.Errorf("rewards buffer is full")
+		case <-receiptsBufFull:
+			log.Info("receipts buffer is full, shutting down")
+			return fmt.Errorf("receipts buffer is full")
```

Comment:
errors should probably use status.Errorf(status.Canceled, "stream buffer is full")

Reply:
good idea